### DISCUSSION TITLE
Migrating to null safety, what is nullable, what is not?

### DISCUSSION THREAD

**Discussion by cervonwong** [simolus3/drift#1220] [Q&A] [answered]
Migrating to null safety, trying to migrate my `Table` classes, but im not sure which types need to be nullable and which can be non-nullable.

Here's an example of a lookup table.

Code after running through the migration tool:
```diff
class Alignments extends Table {
-  IntColumn get id => integer()();
+  IntColumn? get id => integer()();

-  TextColumn get name => text().customConstraint('NOT NULL UNIQUE')();
+  TextColumn? get name => text().customConstraint('NOT NULL UNIQUE')();

  @override
-  Set<Column> get primaryKey => {id};
+  Set<Column?> get primaryKey => {id};
}
```

I get a red squiggly line under `primaryKey` and the error message says `'Alignments.primaryKey' ('Set<Column<dynamic>?> Function()') isn't a valid override of 'Table.primaryKey' ('Set<Column<dynamic>>? Function()').`

Here are my questions:
- Do I need to make my `IntColumn` and `TextColumn` nullable or not?
- How do i define primaryKey in null safety?

Thanks for taking the time to maintain this repo and answering questions (I might ask more questions similar to this as I continue migrating my code.)

**Comment by simolus3** [comment, marked as answer]
>Do I need to make my `IntColumn` and `TextColumn` nullable or not?

No, columns should be non-nullable. Even if a column is marked as `.nullable()`, the column getter shouldn't be marked as nullable in Dart. The columns represent the schema and they're always present, so they should be non-nullable even if the data may be null.

> How do i define primaryKey in null safety?

This is easy when all your columns are marked as non-nullable: Just like you did before :D Your getter should return a `Set<Column>`.

> Thanks for taking the time to maintain this repo and answering questions (I might ask more questions similar to this as I continue migrating my code.)

Thanks, and you're welcome to ask more questions of course :)
